### PR TITLE
Add support for Multiple RNDs in the CHIP Op Certificate Subject DN.

### PR DESCRIPTION
**Pull request by emargolis**
- Introduce new ChipRDN class, which is used by ChipDN class.
  - Added new test cases to verify the functionality.

Other cleanups and changes:
  - Removed AddTrustedKey() function. Currently, spec only defines support
    for trusted anchors in a form of certificates. There is no need to
    support trusted public keys in it's clear form.
  - Removed IsChipX509Attr() and updated IsChipIdX509Attr() functions. Currently,
    only 64-bit CHIP specific OIDs are supported.
  - Removed application defined certificate type override option as not needed for CHIP.

### CODE REVIEW

File: src/credentials/CHIPCert.h [hunk 1]
```diff
@@ -201,11 +198,83 @@ struct ChipDN
...
+ *    A data structure representing a Distinguished Name (DN) in a CHIP certificate.
+ */
+class ChipDN
+{
+public:
+    ChipDN();
+    ~ChipDN();
+
+    /**
+     * @brief Clear DN data.
```
Comment (andy31415) [inline src/credentials/CHIPCert.h:219]: remove comment: saying 'Clear' does a 'Clear' is redundant.

File: src/credentials/CHIPCert.h [hunk 2]
```diff
@@ -201,11 +198,83 @@ struct ChipDN
...
     chip::ASN1::OID mAttrOID;       /**< DN attribute CHIP OID. */
 
-    bool IsEqual(const ChipDN & other) const;
+    bool IsEqual(const ChipRDN & other) const;
     bool IsEmpty() const { return mAttrOID == chip::ASN1::kOID_NotSpecified; }
     void Clear() { mAttrOID = chip::ASN1::kOID_NotSpecified; }
 };
 
+/**
+ *  @struct ChipDN
```
Comment (andy31415) [inline src/credentials/CHIPCert.h:207]: remove `@struct` (this is a class ... also redundant information)

File: src/credentials/CHIPCert.h [hunk 3]
```diff
@@ -201,11 +198,83 @@ struct ChipDN
...
+     *        of its subject DN.
+     *
+     * @param certType  A reference to the certificate type value.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR GetCertType(uint8_t & certType) const;
+
+    /**
+     * @brief Compare two DNs.
```
Comment (andy31415) [inline src/credentials/CHIPCert.h:257]: nit: comment is redundant.

File: src/credentials/CHIPCert.cpp [hunk 4]
```diff
@@ -667,61 +629,148 @@ void ValidationContext::Reset()
...
-                     err = CHIP_ERROR_CERT_USAGE_NOT_ALLOWED);
-
-        // Set the certificate type to CA.
-        cert.mCertType = kCertType_CA;
+        return false;
     }
 
-    // If the certificate subject contains a ChipNodeId attribute set the certificate type to Node.
-    else if (cert.mSubjectDN.mAttrOID == kOID_AttributeType_ChipNodeId)
+    if (IsChipIdX509Attr(mAttrOID))
```
Comment (andy31415) [inline src/credentials/CHIPCert.cpp:639]: should there be a check that mAttrOID equals other.mAttrOID?

File: src/credentials/CHIPCert.cpp [hunk 5]
```diff
@@ -667,61 +629,148 @@ void ValidationContext::Reset()
...
+
+CHIP_ERROR ChipDN::GetCertType(uint8_t & certType) const
+{
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    uint8_t lCertType    = kCertType_NotSpecified;
+    bool fabricIdPresent = false;
+
+    certType = kCertType_NotSpecified;
+
+    for (uint8_t i = 0; i < RDNCount(); i++)
```
Comment (jmartinez-silabs) [inline src/credentials/CHIPCert.cpp:723]: Shouldn't RNDCount() be evaluated before the loop? Calling a for loop at every loop doesn't seem optimal in this case.
Can RDNCount() vary between each loop?

File: src/credentials/CHIPCert.cpp (hunk 4 above)
Comment (andy31415) [reply src/credentials/CHIPCert.cpp:639]: somehow missed it. sorry for the noise.

**Review by andy31415 — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/credentials/CHIPCert.cpp [hunk 6]
```diff
@@ -667,61 +629,150 @@ void ValidationContext::Reset()
...
+}
+
+ChipDN::ChipDN()
+{
+    Clear();
+}
+
+ChipDN::~ChipDN()
+{
+    Clear();
```
Comment (msandstedt) [inline src/credentials/CHIPCert.cpp:657]: Why do we need to manually clear in the destructor?

Likewise, is there a reason we need to manually clear this memory in the constructor?

**Comment by woody-apple** [comment]
@emargolis looks like real build failures here, can you rebase?

File: src/credentials/CHIPCert.cpp (hunk 6 above)
Comment (emargolis) [reply src/credentials/CHIPCert.cpp:657]: removed manual Clear() call from constructor and destructor methods.

**Comment by emargolis** [comment]
> @emargolis looks like real build failures here, can you rebase?

@woody-apple I have rebased and resolved the conflicts.

**Review by bitkis — COMMENTED** [review]
@emargolis,

I understand that for CHIP CA and Software Publisher certificates, the encoding of chip-fabric-id attribute in the DN is optional and ecosystem specific (6.6.6.3. CHIP DN Encoding Rules.)

I'm still not sure how you handle chain validation in case when ICA/Root cert has no chip-fabric-id, but node op cert has one (it would cause a problem, for example, in FindValidCert function.) Am I missing something and such case is not possible? Do we allow CA/Root to contain multiple fabric-id's? It seems in latter case the matching will still fail.

**Comment by emargolis** [comment]
> @emargolis,
> 
> I understand that for CHIP CA and Software Publisher certificates, the encoding of chip-fabric-id attribute in the DN is optional and ecosystem specific (6.6.6.3. CHIP DN Encoding Rules.)
> 
> I'm still not sure how you handle chain validation in case when ICA/Root cert has no chip-fabric-id, but node op cert has one (it would cause a problem, for example, in FindValidCert function.) Am I missing something and such case is not possible? Do we allow CA/Root to contain multiple fabric-id's? It seems in latter case the matching will still fail.

@bitkis 

Each certificate has Subject and Issuer DNs. Just as an example from my test vectors in this PR:

```
For Node01_01 Certificate:
Subject: 1.3.6.1.4.1.37244.1.1 = DEDEDEDE00010001, 1.3.6.1.4.1.37244.1.4 = FAB000000000001D
Issuer: 1.3.6.1.4.1.37244.1.2 = CACACACA00000003

For ICA01 Certificate:
Subject: 1.3.6.1.4.1.37244.1.2 = CACACACA00000003
Issuer: 1.3.6.1.4.1.37244.1.2 = CACACACA00000001

For Root01 Certificate:
Subject: 1.3.6.1.4.1.37244.1.2 = CACACACA00000001
Issuer: 1.3.6.1.4.1.37244.1.2 = CACACACA00000001
```

The IssuerDN of the Node01_01 certificate will be matched with the SubjectDN of the ICA01 certificate.
The IssuerDN of the ICA01 certificate will be matched with the SubjectDN of the Root01 certificate.

**Review by yufengwangca — APPROVED** [review]

**Comment by yufengwangca** [comment]
@woody-apple ?

**Review by bitkis — APPROVED** [review]

**Review by mspang — APPROVED** [review]

File: src/lib/core/CHIPConfig.h [hunk 7]
```diff
@@ -1463,6 +1463,18 @@
...
+/**
+ *  @def CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES
+ *
+ *  @brief
+ *    The maximum number of Relative Distinguished Name (RDN) attributes
+ *    supported by the CHIP Certificate Distinguished Names (DN).
+ *
+ */
+#ifndef CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES
+#define CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES 3
```
Comment (mspang) [inline src/lib/core/CHIPConfig.h:1475]: Should this be 5 ?

Comment (emargolis) [reply src/lib/core/CHIPConfig.h]: updated in my latest commit, which address new Authentication Tag attributes

**Comment by emargolis** [comment]
My latest commit addresses latest DN attributes changes in the spec and aligns implementation to the spec. Specifically: 
-- Added support for 5 DN attributes in the DN.
-- Added support for Root and Authentication Tag attributes
-- Updated CHIP OIDs definision accoring to the spec
-- Added new functions IsChip64bitDNAttr() IsChip32bitDNAttr() and IsChipDNAttr()

**Comment by bitkis** [comment]
@emargolis, to make review easier, please squash 9abdbf9 and 9abdbf9

**Review by msandstedt — APPROVED** [review]

**Pull request merged**